src/interface/blas1_interface.hpp

```diff
@@ -246,6 +246,108 @@ typename sb_handle_t::event_t _asum_impl(
+ * @param _incx Stride of vector x (i.e. measured in elements of _vx)
+ * @param _rs output variable
+ * @param _nWG no. of work groups required for the reduction
+ * @param _dependencies Vector of events
+ */
+
+template <int localSize, int localMemSize, bool is_max, bool single,
+          typename sb_handle_t, typename container_0_t, typename container_1_t,
+          typename index_t, typename increment_t>
+typename sb_handle_t::event_t _iamax_iamin_impl(
+    sb_handle_t &sb_handle, index_t _N, container_0_t _vx, increment_t _incx,
+    container_1_t _rs, const index_t _nWG,
+    const typename sb_handle_t::event_t &_dependencies) {
+  typename VectorViewType<container_0_t, index_t, increment_t>::type vx =
+      make_vector_view(_vx, _incx, _N);
+  auto rs = make_vector_view<index_t>(_rs, static_cast<increment_t>(1),
+                                      static_cast<index_t>(1));
+  auto tupOp = make_tuple_op(vx);
+  typename sb_handle_t::event_t ret;
+  if constexpr (single) {
+    auto op = make_index_max_min<is_max, false>(rs, tupOp);
+    if constexpr (localMemSize == 0) {
+      auto q = sb_handle.get_queue();
+      const index_t sg_size = static_cast<index_t>(
+          q.get_device()
+              .template get_info<sycl::info::device::sub_group_sizes>()[0]);
+      ret = sb_handle.execute(op, sg_size, sg_size, _dependencies);
+    } else {
+      ret =
+          sb_handle.execute(op, static_cast<index_t>(localSize),
+                            _nWG * static_cast<index_t>(localSize),
+                            static_cast<index_t>(localMemSize), _dependencies);
+    }
+  } else {
+    using scalar_t = typename ValueType<container_0_t>::type;
+    using tuple_t = IndexValueTuple<index_t, scalar_t>;
+    constexpr bool is_usm = std::is_pointer<container_0_t>::value;
+    auto q = sb_handle.get_queue();
+    const index_t sg_size = static_cast<index_t>(
+        q.get_device()
+            .template get_info<sycl::info::device::sub_group_sizes>()[0]);
+    const index_t memory_size =
+        localMemSize == 0 ? _nWG * (static_cast<index_t>(localSize) / sg_size)
+                          : _nWG;
+    auto gpu_res = blas::helper::allocate < is_usm ? helper::AllocType::usm
+                                                   : helper::AllocType::buffer,
+         tuple_t > (memory_size, q);
+    auto gpu_res_vec =
+        make_vector_view(gpu_res, static_cast<increment_t>(1), memory_size);
+    auto step0 = make_index_max_min<is_max, true>(gpu_res_vec, tupOp);
+    auto step1 = make_index_max_min<is_max, false>(rs, gpu_res_vec);
+    if constexpr (localMemSize == 0) {
+      const scalar_t val = is_max ? std::numeric_limits<scalar_t>::min()
+                                  : std::numeric_limits<scalar_t>::max();
+      const index_t idx = std::numeric_limits<index_t>::max();
+      const tuple_t init{idx, val};
+      ret = typename sb_handle_t::event_t{
+          helper::fill(q, gpu_res, init, memory_size, _dependencies)};
```

Comment:
I'm missing the reason why we need to initilize `gpu_res` just in case of `localMemSize == 0`.

Reply:
When `localMemSize == 0`, the implementation assumes that it is going to be executed on the CPU. The size of `gpu_res` is dependant on the `sg_size` variable since we want every `sub_group` to write one intermediate output to global memory. Since the `sub_group` size (to be used by the kernel) is not known at compile time, I allocated a larger memory using the smallest `sub_group` size supported by the CPU. This means that there could be a case where the actual `sub_group` size chosen by the compiler is larger than the one used to calculate the size of `gpu_res`, and this would mean that there could be locations inside `gpu_res` with garbage value that would effect the kernel output in the final reduction. That's why we have to initialize the intermediate memory for the CPU case only. For GPUs, every work group writes 1 value of data to global memory, and since the no. of work groups are known, we don't need to initialize the memory beforehand.

Reply:
e.g. on the device I tested, smallest sub_group size supported was `4` whereas the sub_group size used by the kernel was `16`.

Reply:
Ok, fine. Please, could you add some comments for explaining it and change the variable names accordingly? E.g., `sg_size -> min_sg_size`.